keras_cv/models/object_detection/faster_rcnn/faster_rcnn.py

```diff
@@ -459,41 +318,46 @@ def compile(
         }
         super().compile(loss=losses, **kwargs)
 
-    def compute_loss(self, images, boxes, classes, training):
+    def compute_loss(self, x, y, y_pred, sample_weight, **kwargs):
+        images = x
+        boxes = y["boxes"]
+        if len(y["classes"].shape) != 2:
+            raise ValueError(
+                "Expected 'classes' to be a tf.Tensor of rank 2. "
+                f"Got y['classes'].shape={y['classes'].shape}."
+            )
+        # TODO(tanzhenyu): remove this hack and perform broadcasting elsewhere
+        classes = ops.expand_dims(y["classes"], axis=-1)
+
         local_batch = images.get_shape().as_list()[0]
-        if tf.distribute.has_strategy():
-            num_sync = tf.distribute.get_strategy().num_replicas_in_sync
-        else:
-            num_sync = 1
-        global_batch = local_batch * num_sync
         anchors = self.anchor_generator(image_shape=tuple(images[0].shape))
         (
             rpn_box_targets,
             rpn_box_weights,
             rpn_cls_targets,
             rpn_cls_weights,
         ) = self.rpn_labeler(
-            tf.concat(tf.nest.flatten(anchors), axis=0), boxes, classes
+            ops.concatenate(tree.flatten(anchors), axis=0), boxes, classes
         )
         rpn_box_weights /= (
-            self.rpn_labeler.samples_per_image * global_batch * 0.25
+            self.rpn_labeler.samples_per_image * local_batch * 0.25
         )
-        rpn_cls_weights /= self.rpn_labeler.samples_per_image * global_batch
+        rpn_cls_weights /= self.rpn_labeler.samples_per_image * local_batch
         rois, feature_map, rpn_box_pred, rpn_cls_pred = self._call_rpn(
-            images, anchors, training=training
+            images, anchors, training=kwargs["training"]
```

Comment:
@ianstenbit do you have any pointers on how to pass the `training` flag?

Reply:
First let's verify that we actually need to -- is this meaningfully used anywhere downstream? I can't see it being used anywhere.

Reply:
The `training` parameter is required for the `_call_rpn` and `_call_rcnn`  methods. They both use it to call the `feature_extractor`, `feature_pyramid`, `rpn_head`, `roi_generator`, and the `rcnn_head` layers. WDYT?

Reply:
Most of those downstream layers don't actually _use_ this param -- looks like only the `roi_generator` does. I am trying to figure out the best way to make this work correctly.

Reply:
Here's a solution to try:
It looks like all of these values are _already_ being computed in `call`, and ignored.
Instead of duplicating these calls to `_call_rpn` and `_call_rcnn`, can we simply pass the results of those methods to the `compute_loss` function by returning them in `call`?

(Then they'll show up in the `y_pred` parameter in `compute_loss` -- let's structure them as a dictionary). This is a more elegant structure anyway as it reduces the duplication of the code.

Reply:
edit: We can actually just skip passing this `training` argument explicitly, because it will be automatically inferred based on call context when the layer is called. So we can just drop any reference to `training` from the `compute_loss` function.